flowmachine/flowmachine/features/location/majority_location.py

```diff
@@ -0,0 +1,35 @@
+from typing import List
+
+from flowmachine.core import Query
+from flowmachine.core.mixins import GeoDataMixin
+from flowmachine.features.utilities.subscriber_locations import BaseLocation
+
+
+class MajorityLocation(BaseLocation, Query):
+    def __init__(self, subscriber_subset: Query, weight_column):
+        if weight_column not in subscriber_subset.column_names:
+            raise ValueError("weight_column must exist in subscriber_subset")
+        self.subscriber_subset = subscriber_subset
+        self.weight_column = weight_column
+        super().__init__()
+
+    @property
+    def column_names(self) -> List[str]:
+        # This should be a reduction, so no new columns
+        return ["subscriber"] + self.subscriber_subset.spatial_unit.location_id_columns
+
+    def _make_query(self):
+        loc_id = ",".join(self.subscriber_subset.spatial_unit.location_id_columns)
+        sql = f"""
+WITH subscriber_subset AS (
```

Comment:
So interesting to note that until quite recently (pg 12) CTEs were an optimization fence, and often to be avoided in favour of a subquery. Now, what it'll do is rewrite it as a subquery for you prior to planning in a case like this where there's only one use of the subquery - see https://www.postgresql.org/docs/12/queries-with.html for more (when we wrote very early versions of fk, we didn't know about the CTE behaviour, so we used it liberally, and then removed _all_ uses of it cos it generally led to dreadfully slow query plans).